src/flamegraph/color/palette_map.rs

```diff
@@ -1,84 +1,136 @@
+        let mut map = HashMap::default();
 
-                let rgb_color = parse_rgb_string(color)
-                    .ok_or_else(|| quick_xml::Error::UnexpectedToken(color.to_string()))?;
-                map.insert(Cow::from(name.to_string()), rgb_color);
-            }
+        for line in reader.lines() {
+            let line = line?;
+            let (name, color) = parse_line(&line)?;
+            map.insert(name.to_string(), color);
         }
 
         Ok(PaletteMap(map))
     }
 
-    pub fn save(self, file: &str) -> Result<(), io::Error> {
-        let mut file = OpenOptions::new().write(true).create(true).open(file)?;
-        let mut entries = self.0.into_iter().collect::<Vec<_>>();
+    /// Writes the palette map using the given writer.
+    /// The output content will follow the same format described in [from_stream()]
+    /// The name/color pairs will be sorted by name in lexicographic order.
+    pub fn to_stream(&self, writer: &mut dyn io::Write) -> io::Result<()> {
+        let mut entries = self.0.iter().collect::<Vec<_>>();
         // We sort the palette because the Perl implementation does.
         entries.sort_unstable();
 
         for (name, color) in entries {
-            file.write_all(
-                format!("{}->rgb({},{},{})\n", name, color.0, color.1, color.2).as_bytes(),
+            writer.write_all(
+                format!("{}->rgb({},{},{})\n", name, color.r, color.g, color.b).as_bytes(),
             )?
         }
 
         Ok(())
     }
 
-    pub fn find_color_for<F: FnMut(&'a str) -> (u8, u8, u8)>(
+    /// Utility function to load a palette map from a file.
+    ///
+    /// The file content should follow the format described in [from_stream()].
+    ///
+    /// If the file does not exist, an empty palette map is returned.
+    pub fn load_from_file_or_empty(path: &dyn AsRef<Path>) -> io::Result<Self> {
+        // If the file does not exist, it is probably the first call to flamegraph with a consistent
+        // palette: there is nothing to load.
+        if path.as_ref().exists() {
+            let file = File::open(path)?;
+            let mut reader = BufReader::new(file);
+            PaletteMap::from_stream(&mut reader)
+        } else {
+            Ok(PaletteMap::default())
+        }
+    }
+
+    /// Utility function to save a palette map to a file.
+    ///
+    /// The file content will follow the format described in [from_stream()].
+    pub fn save_to_file(&self, path: &dyn AsRef<Path>) -> io::Result<()> {
+        let mut file = OpenOptions::new().write(true).create(true).open(path)?;
+        self.to_stream(&mut file)
+    }
+
+    /// Returns the color value corresponding to the given function name if it is present.
+    /// Otherwise compute the color, and insert the new function name/color in the map.
+    pub(crate) fn find_color_for<S: ToString, F: FnMut(&S) -> Color>(
         &mut self,
-        name: &'a str,
+        name: &S,
```

Comment:
Hmm, this doesn't seem right. This will cause us to double-allocate every string we look up colors for, no? I wonder if it'll be better here to take `&str` and do the `.get` and `.insert` calls separately?

Reply:
[Done](https://github.com/jonhoo/inferno/pull/77/commits/14ddd2402e973a7b993511e49488ab75438bc0ad)